File: JavaScript/packages/recognizers-number-with-unit/src/numberWithUnit/numberWithUnitRecognizer.ts
import { IModel, ModelResult, Recognizer, IExtractor, IParser } from "@microsoft/recognizers-text";
import { Culture } from "@microsoft/recognizers-text-number";
import { CurrencyModel, TemperatureModel, DimensionModel, AgeModel } from "./models";
import { NumberWithUnitExtractor, BaseMergedUnitExtractor } from "./extractors";
import { NumberWithUnitParser, BaseMergedUnitParser } from "./parsers";
import { EnglishCurrencyExtractorConfiguration, EnglishCurrencyParserConfiguration } from "./english/currency";
import { EnglishTemperatureExtractorConfiguration, EnglishTemperatureParserConfiguration } from "./english/temperature";
import { EnglishDimensionExtractorConfiguration, EnglishDimensionParserConfiguration } from "./english/dimension";
import { EnglishAgeExtractorConfiguration, EnglishAgeParserConfiguration } from "./english/age";
import { SpanishCurrencyExtractorConfiguration, SpanishCurrencyParserConfiguration } from "./spanish/currency";
import { SpanishTemperatureExtractorConfiguration, SpanishTemperatureParserConfiguration } from "./spanish/temperature";
import { SpanishDimensionExtractorConfiguration, SpanishDimensionParserConfiguration } from "./spanish/dimension";
import { SpanishAgeExtractorConfiguration, SpanishAgeParserConfiguration } from "./spanish/age";
import { PortugueseCurrencyExtractorConfiguration, PortugueseCurrencyParserConfiguration } from "./portuguese/currency";
import { PortugueseTemperatureExtractorConfiguration, PortugueseTemperatureParserConfiguration } from "./portuguese/temperature";
import { PortugueseDimensionExtractorConfiguration, PortugueseDimensionParserConfiguration } from "./portuguese/dimension";
import { PortugueseAgeExtractorConfiguration, PortugueseAgeParserConfiguration } from "./portuguese/age";
import { ChineseCurrencyExtractorConfiguration, ChineseCurrencyParserConfiguration } from "./chinese/currency";
import { ChineseTemperatureExtractorConfiguration, ChineseTemperatureParserConfiguration } from "./chinese/temperature";
import { ChineseDimensionExtractorConfiguration, ChineseDimensionParserConfiguration } from "./chinese/dimension";
import { ChineseAgeExtractorConfiguration, ChineseAgeParserConfiguration } from "./chinese/age";
import { FrenchCurrencyExtractorConfiguration, FrenchCurrencyParserConfiguration } from "./french/currency";
import { FrenchTemperatureExtractorConfiguration, FrenchTemperatureParserConfiguration } from "./french/temperature";
import { FrenchDimensionExtractorConfiguration, FrenchDimensionParserConfiguration } from "./french/dimension";
import { FrenchAgeExtractorConfiguration, FrenchAgeParserConfiguration } from "./french/age";

export enum NumberWithUnitOptions {
    None = 0,
}

export function recognizeCurrency(query: string, culture: string, options: NumberWithUnitOptions = NumberWithUnitOptions.None, fallbackToDefaultCulture: boolean = true): Array<ModelResult> {
    return recognizeByModel(recognizer => recognizer.getCurrencyModel(culture, fallbackToDefaultCulture), query, culture, options);
}

export function recognizeTemperature(query: string, culture: string, options: NumberWithUnitOptions = NumberWithUnitOptions.None, fallbackToDefaultCulture: boolean = true): Array<ModelResult> {
    return recognizeByModel(recognizer => recognizer.getTemperatureModel(culture, fallbackToDefaultCulture), query, culture, options);
}

export function recognizeDimension(query: string, culture: string, options: NumberWithUnitOptions = NumberWithUnitOptions.None, fallbackToDefaultCulture: boolean = true): Array<ModelResult> {
    return recognizeByModel(recognizer => recognizer.getDimensionModel(culture, fallbackToDefaultCulture), query, culture, options);
}

export function recognizeAge(query: string, culture: string, options: NumberWithUnitOptions = NumberWithUnitOptions.None, fallbackToDefaultCulture: boolean = true): Array<ModelResult> {
    return recognizeByModel(recognizer => recognizer.getAgeModel(culture, fallbackToDefaultCulture), query, culture, options);
}

function recognizeByModel(getModelFunc: (n: NumberWithUnitRecognizer) => IModel, query: string, culture: string, options: NumberWithUnitOptions): Array<ModelResult> {
    let recognizer = new NumberWithUnitRecognizer(culture, options);
    let model = getModelFunc(recognizer);
    return model.parse(query);
}

export default class NumberWithUnitRecognizer extends Recognizer<NumberWithUnitOptions> {
    constructor(culture: string, options: NumberWithUnitOptions = NumberWithUnitOptions.None, lazyInitialization: boolean = false) {
        super(culture, options, lazyInitialization);
    }

    protected InitializeConfiguration() {
        //#region English
        this.registerModel("CurrencyModel", Culture.English, (options) => new CurrencyModel(new Map<IExtractor, IParser>([
            [new BaseMergedUnitExtractor(new EnglishCurrencyExtractorConfiguration()), new BaseMergedUnitParser(new EnglishCurrencyParserConfiguration())]
        ])));
        this.registerModel("TemperatureModel", Culture.English, (options) => new TemperatureModel(new Map<IExtractor, IParser>([
            [new NumberWithUnitExtractor(new EnglishTemperatureExtractorConfiguration()), new NumberWithUnitParser(new EnglishTemperatureParserConfiguration())]
        ])));
        this.registerModel("DimensionModel", Culture.English,(options) =>  new DimensionModel(new Map<IExtractor, IParser>([
            [new NumberWithUnitExtractor(new EnglishDimensionExtractorConfiguration()), new NumberWithUnitParser(new EnglishDimensionParserConfiguration())]
        ])));
        this.registerModel("AgeModel", Culture.English, (options) => new AgeModel(new Map<IExtractor, IParser>([
            [new NumberWithUnitExtractor(new EnglishAgeExtractorConfiguration()), new NumberWithUnitParser(new EnglishAgeParserConfiguration())]
        ])));
        //#endregion

        //#region Spanish
        this.registerModel("CurrencyModel", Culture.Spanish, (options) => new CurrencyModel(new Map<IExtractor, IParser>([
            [new NumberWithUnitExtractor(new SpanishCurrencyExtractorConfiguration()), new NumberWithUnitParser(new SpanishCurrencyParserConfiguration())]
        ])));
        this.registerModel("TemperatureModel", Culture.Spanish, (options) => new TemperatureModel(new Map<IExtractor, IParser>([
            [new NumberWithUnitExtractor(new SpanishTemperatureExtractorConfiguration()), new NumberWithUnitParser(new SpanishTemperatureParserConfiguration())]
        ])));
        this.registerModel("DimensionModel", Culture.Spanish, (options) => new DimensionModel(new Map<IExtractor, IParser>([
            [new NumberWithUnitExtractor(new SpanishDimensionExtractorConfiguration()), new NumberWithUnitParser(new SpanishDimensionParserConfiguration())]
        ])));
        this.registerModel("AgeModel", Culture.Spanish, (options) => new AgeModel(new Map<IExtractor, IParser>([
            [new NumberWithUnitExtractor(new SpanishAgeExtractorConfiguration()), new NumberWithUnitParser(new SpanishAgeParserConfiguration())]
        ])));
        //#endregion

        //#region Portuguese
        this.registerModel("CurrencyModel", Culture.Portuguese, (options) => new CurrencyModel(new Map<IExtractor, IParser>([
            [new NumberWithUnitExtractor(new PortugueseCurrencyExtractorConfiguration()), new NumberWithUnitParser(new PortugueseCurrencyParserConfiguration())]
        ])));
        this.registerModel("TemperatureModel", Culture.Portuguese, (options) => new TemperatureModel(new Map<IExtractor, IParser>([
            [new NumberWithUnitExtractor(new PortugueseTemperatureExtractorConfiguration()), new NumberWithUnitParser(new PortugueseTemperatureParserConfiguration())]
        ])));
        this.registerModel("DimensionModel", Culture.Portuguese, (options) => new DimensionModel(new Map<IExtractor, IParser>([
            [new NumberWithUnitExtractor(new PortugueseDimensionExtractorConfiguration()), new NumberWithUnitParser(new PortugueseDimensionParserConfiguration())]
        ])));
        this.registerModel("AgeModel", Culture.Portuguese, (options) => new AgeModel(new Map<IExtractor, IParser>([
            [new NumberWithUnitExtractor(new PortugueseAgeExtractorConfiguration()), new NumberWithUnitParser(new PortugueseAgeParserConfiguration())]
        ])));
        //#endregion

        //#region Chinese
        this.registerModel("CurrencyModel", Culture.Chinese, (options) => new CurrencyModel(new Map<IExtractor, IParser>([
            [new BaseMergedUnitExtractor(new ChineseCurrencyExtractorConfiguration()), new BaseMergedUnitParser(new ChineseCurrencyParserConfiguration())],
            [new NumberWithUnitExtractor(new EnglishCurrencyExtractorConfiguration()), new NumberWithUnitParser(new EnglishCurrencyParserConfiguration())]
        ])));
        this.registerModel("TemperatureModel", Culture.Chinese, (options) => new TemperatureModel(new Map<IExtractor, IParser>([
            [new NumberWithUnitExtractor(new ChineseTemperatureExtractorConfiguration()), new NumberWithUnitParser(new ChineseTemperatureParserConfiguration())],
            [new NumberWithUnitExtractor(new EnglishTemperatureExtractorConfiguration()), new NumberWithUnitParser(new EnglishTemperatureParserConfiguration())]
        ])));
        this.registerModel("DimensionModel", Culture.Chinese, (options) => new DimensionModel(new Map<IExtractor, IParser>([
            [new NumberWithUnitExtractor(new ChineseDimensionExtractorConfiguration()), new NumberWithUnitParser(new ChineseDimensionParserConfiguration())],
            [new NumberWithUnitExtractor(new EnglishDimensionExtractorConfiguration()), new NumberWithUnitParser(new EnglishDimensionParserConfiguration())]
        ])));
        this.registerModel("AgeModel", Culture.Chinese, (options) => new AgeModel(new Map<IExtractor, IParser>([
            [new NumberWithUnitExtractor(new ChineseAgeExtractorConfiguration()), new NumberWithUnitParser(new ChineseAgeParserConfiguration())],
            [new NumberWithUnitExtractor(new EnglishAgeExtractorConfiguration()), new NumberWithUnitParser(new EnglishAgeParserConfiguration())]
        ])));
        //#endregion

        //#region French
        this.registerModel("CurrencyModel", Culture.French, (options) => new CurrencyModel(new Map<IExtractor, IParser>([
            [new NumberWithUnitExtractor(new FrenchCurrencyExtractorConfiguration()), new NumberWithUnitParser(new FrenchCurrencyParserConfiguration())]
        ])));
        this.registerModel("TemperatureModel", Culture.French, (options) => new TemperatureModel(new Map<IExtractor, IParser>([
            [new NumberWithUnitExtractor(new FrenchTemperatureExtractorConfiguration()), new NumberWithUnitParser(new FrenchTemperatureParserConfiguration())]
        ])));
        this.registerModel("DimensionModel", Culture.French, (options) => new DimensionModel(new Map<IExtractor, IParser>([
            [new NumberWithUnitExtractor(new FrenchDimensionExtractorConfiguration()), new NumberWithUnitParser(new FrenchDimensionParserConfiguration())]
        ])));
        this.registerModel("AgeModel", Culture.French, (options) => new AgeModel(new Map<IExtractor, IParser>([
            [new NumberWithUnitExtractor(new FrenchAgeExtractorConfiguration()), new NumberWithUnitParser(new FrenchAgeParserConfiguration())]
        ])));
        //#endregion
    }

    protected IsValidOptions(options: number): boolean {
        return options >= 0 && options <= NumberWithUnitOptions.None
    }

    getCurrencyModel(culture: string = null, fallbackToDefaultCulture: boolean = true): IModel {
        return this.getModel("CurrencyModel", culture, fallbackToDefaultCulture);
    }

    getTemperatureModel(culture: string = null, fallbackToDefaultCulture: boolean = true): IModel {
        return this.getModel("TemperatureModel", culture, fallbackToDefaultCulture);
    }

    getDimensionModel(culture: string = null, fallbackToDefaultCulture: boolean = true): IModel {
        return this.getModel("DimensionModel", culture, fallbackToDefaultCulture);
    }

    getAgeModel(culture: string = null, fallbackToDefaultCulture: boolean = true): IModel {
        return this.getModel("AgeModel", culture, fallbackToDefaultCulture);
    }
}

File: Python/libraries/recognizers-number-with-unit/recognizers_number_with_unit/number_with_unit/number_with_unit_recognizer.py
#  Copyright (c) Microsoft Corporation. All rights reserved.
#  Licensed under the MIT License.

from enum import IntFlag
from typing import List
from recognizers_text import Culture, Recognizer
from recognizers_text.model import Model, ModelResult
from recognizers_number.culture import CultureInfo
from .dutch.extractors import DutchCurrencyExtractorConfiguration, DutchTemperatureExtractorConfiguration, \
    DutchDimensionExtractorConfiguration, DutchAgeExtractorConfiguration
from .dutch.parsers import DutchCurrencyParserConfiguration, DutchTemperatureParserConfiguration, \
    DutchDimensionParserConfiguration, DutchAgeParserConfiguration
from .models import CurrencyModel, TemperatureModel, DimensionModel, AgeModel, ExtractorParserModel
from .extractors import NumberWithUnitExtractor, BaseMergedUnitExtractor
from .parsers import NumberWithUnitParser, BaseMergedUnitParser
from .english.extractors import (EnglishCurrencyExtractorConfiguration,
                                 EnglishTemperatureExtractorConfiguration,
                                 EnglishDimensionExtractorConfiguration,
                                 EnglishAgeExtractorConfiguration)
from .english.parsers import (EnglishCurrencyParserConfiguration,
                              EnglishTemperatureParserConfiguration,
                              EnglishDimensionParserConfiguration,
                              EnglishAgeParserConfiguration)
from .chinese.extractors import (ChineseCurrencyExtractorConfiguration,
                                 ChineseTemperatureExtractorConfiguration,
                                 ChineseDimensionExtractorConfiguration,
                                 ChineseAgeExtractorConfiguration)
from .chinese.parsers import (ChineseCurrencyParserConfiguration,
                              ChineseTemperatureParserConfiguration,
                              ChineseDimensionParserConfiguration,
                              ChineseAgeParserConfiguration)
from .dutch.extractors import (DutchCurrencyExtractorConfiguration,
                               DutchTemperatureExtractorConfiguration,
                               DutchDimensionExtractorConfiguration,
                               DutchAgeExtractorConfiguration)
from .dutch.parsers import (DutchCurrencyParserConfiguration,
                            DutchTemperatureParserConfiguration,
                            DutchDimensionParserConfiguration,
                            DutchAgeParserConfiguration)
from .spanish.extractors import (SpanishCurrencyExtractorConfiguration,
                                 SpanishTemperatureExtractorConfiguration,
                                 SpanishDimensionExtractorConfiguration,
                                 SpanishAgeExtractorConfiguration)
from .spanish.parsers import (SpanishCurrencyParserConfiguration,
                              SpanishTemperatureParserConfiguration,
                              SpanishDimensionParserConfiguration,
                              SpanishAgeParserConfiguration)
from .french.extractors import (FrenchCurrencyExtractorConfiguration,
                                FrenchTemperatureExtractorConfiguration,
                                FrenchDimensionExtractorConfiguration,
                                FrenchAgeExtractorConfiguration)
from .french.parsers import (FrenchCurrencyParserConfiguration,
                             FrenchTemperatureParserConfiguration,
                             FrenchDimensionParserConfiguration,
                             FrenchAgeParserConfiguration)
from .portuguese.extractors import (PortugueseCurrencyExtractorConfiguration,
                                    PortugueseTemperatureExtractorConfiguration,
                                    PortugueseDimensionExtractorConfiguration,
                                    PortugueseAgeExtractorConfiguration)
from .portuguese.parsers import (PortugueseCurrencyParserConfiguration,
                                 PortugueseTemperatureParserConfiguration,
                                 PortugueseDimensionParserConfiguration,
                                 PortugueseAgeParserConfiguration)
from .german.extractors import GermanCurrencyExtractorConfiguration
from .german.parsers import GermanCurrencyParserConfiguration
from .italian.extractors import ItalianCurrencyExtractorConfiguration
from .italian.parsers import ItalianCurrencyParserConfiguration
from .japanese.extractors import JapaneseCurrencyExtractorConfiguration
from .japanese.parsers import JapaneseCurrencyParserConfiguration


class NumberWithUnitOptions(IntFlag):
    NONE = 0


class NumberWithUnitRecognizer(Recognizer[NumberWithUnitOptions]):
    def __init__(self, target_culture: str = None,
                 options: NumberWithUnitOptions = NumberWithUnitOptions.NONE,
                 lazy_initialization: bool = True):
        if options < NumberWithUnitOptions.NONE or options > NumberWithUnitOptions.NONE:
            raise ValueError()
        super().__init__(target_culture, options, lazy_initialization)

    def initialize_configuration(self):
        # region English
        self.register_model('CurrencyModel', Culture.English, lambda options: CurrencyModel(
            [ExtractorParserModel(BaseMergedUnitExtractor(EnglishCurrencyExtractorConfiguration(
            )), BaseMergedUnitParser(EnglishCurrencyParserConfiguration()))]
        ))
        self.register_model('TemperatureModel', Culture.English, lambda options: TemperatureModel(
            [ExtractorParserModel(NumberWithUnitExtractor(EnglishTemperatureExtractorConfiguration(
            )), NumberWithUnitParser(EnglishTemperatureParserConfiguration()))]
        ))
        self.register_model('DimensionModel', Culture.English, lambda options: DimensionModel(
            [ExtractorParserModel(NumberWithUnitExtractor(EnglishDimensionExtractorConfiguration(
            )), NumberWithUnitParser(EnglishDimensionParserConfiguration()))]
        ))
        self.register_model('AgeModel', Culture.English, lambda options: AgeModel(
            [ExtractorParserModel(NumberWithUnitExtractor(EnglishAgeExtractorConfiguration(
            )), NumberWithUnitParser(EnglishAgeParserConfiguration()))]
        ))
        # endregion

        # region Chinese
        self.register_model('CurrencyModel', Culture.Chinese, lambda options: CurrencyModel([
            ExtractorParserModel(
                BaseMergedUnitExtractor(
                    ChineseCurrencyExtractorConfiguration()),
                BaseMergedUnitParser(ChineseCurrencyParserConfiguration())),
            ExtractorParserModel(
                NumberWithUnitExtractor(
                    EnglishCurrencyExtractorConfiguration()),
                NumberWithUnitParser(EnglishCurrencyParserConfiguration()))
        ]))
        self.register_model('TemperatureModel', Culture.Chinese, lambda options: TemperatureModel([
            ExtractorParserModel(
                NumberWithUnitExtractor(
                    ChineseTemperatureExtractorConfiguration()),
                NumberWithUnitParser(ChineseTemperatureParserConfiguration())),
            ExtractorParserModel(
                NumberWithUnitExtractor(
                    EnglishTemperatureExtractorConfiguration()),
                NumberWithUnitParser(EnglishTemperatureParserConfiguration()))
        ]))
        self.register_model('DimensionModel', Culture.Chinese, lambda options: DimensionModel([
            ExtractorParserModel(
                NumberWithUnitExtractor(
                    ChineseDimensionExtractorConfiguration()),
                NumberWithUnitParser(ChineseDimensionParserConfiguration())),
            ExtractorParserModel(
                NumberWithUnitExtractor(
                    EnglishDimensionExtractorConfiguration()),
                NumberWithUnitParser(EnglishDimensionParserConfiguration()))
        ]))
        self.register_model('AgeModel', Culture.Chinese, lambda options: AgeModel([
            ExtractorParserModel(
                NumberWithUnitExtractor(ChineseAgeExtractorConfiguration()),
                NumberWithUnitParser(ChineseAgeParserConfiguration())),
            ExtractorParserModel(
                NumberWithUnitExtractor(EnglishAgeExtractorConfiguration()),
                NumberWithUnitParser(EnglishAgeParserConfiguration()))
        ]))
        # endregion

        # region Dutch
        self.register_model('CurrencyModel', Culture.Dutch, lambda options: CurrencyModel(
            [ExtractorParserModel(BaseMergedUnitExtractor(DutchCurrencyExtractorConfiguration(
            )), BaseMergedUnitParser(DutchCurrencyParserConfiguration()))]
        ))
        self.register_model('TemperatureModel', Culture.Dutch, lambda options: TemperatureModel([
            ExtractorParserModel(
                NumberWithUnitExtractor(
                    DutchTemperatureExtractorConfiguration()),
                NumberWithUnitParser(DutchTemperatureParserConfiguration()))
        ]))
        self.register_model('DimensionModel', Culture.Dutch, lambda options: DimensionModel([
            ExtractorParserModel(
                NumberWithUnitExtractor(
                    DutchDimensionExtractorConfiguration()),
                NumberWithUnitParser(DutchDimensionParserConfiguration()))
        ]))
        self.register_model('AgeModel', Culture.Dutch, lambda options: AgeModel([
            ExtractorParserModel(
                NumberWithUnitExtractor(DutchAgeExtractorConfiguration()),
                NumberWithUnitParser(DutchAgeParserConfiguration()))
        ]))
        # endregion

        # region French
        self.register_model('CurrencyModel', Culture.French, lambda options: CurrencyModel(
            [ExtractorParserModel(BaseMergedUnitExtractor(FrenchCurrencyExtractorConfiguration(
            )), BaseMergedUnitParser(FrenchCurrencyParserConfiguration()))]
        ))
        self.register_model('TemperatureModel', Culture.French, lambda options: TemperatureModel([
            ExtractorParserModel(
                NumberWithUnitExtractor(
                    FrenchTemperatureExtractorConfiguration()),
                NumberWithUnitParser(FrenchTemperatureParserConfiguration()))
        ]))
        self.register_model('DimensionModel', Culture.French, lambda options: DimensionModel([
            ExtractorParserModel(
                NumberWithUnitExtractor(
                    FrenchDimensionExtractorConfiguration()),
                NumberWithUnitParser(FrenchDimensionParserConfiguration()))
        ]))
        self.register_model('AgeModel', Culture.French, lambda options: AgeModel([
            ExtractorParserModel(
                NumberWithUnitExtractor(FrenchAgeExtractorConfiguration()),
                NumberWithUnitParser(FrenchAgeParserConfiguration()))
        ]))
        # endregion

        # region Portuguese
        self.register_model('CurrencyModel', Culture.Portuguese, lambda options: CurrencyModel(
            [ExtractorParserModel(BaseMergedUnitExtractor(PortugueseCurrencyExtractorConfiguration(
            )), BaseMergedUnitParser(PortugueseCurrencyParserConfiguration()))]
        ))
        self.register_model('TemperatureModel', Culture.Portuguese, lambda options: TemperatureModel([
            ExtractorParserModel(
                NumberWithUnitExtractor(
                    PortugueseTemperatureExtractorConfiguration()),
                NumberWithUnitParser(PortugueseTemperatureParserConfiguration()))
        ]))
        self.register_model('DimensionModel', Culture.Portuguese, lambda options: DimensionModel([
            ExtractorParserModel(
                NumberWithUnitExtractor(
                    PortugueseDimensionExtractorConfiguration()),
                NumberWithUnitParser(PortugueseDimensionParserConfiguration()))
        ]))
        self.register_model('AgeModel', Culture.Portuguese, lambda options: AgeModel([
            ExtractorParserModel(
                NumberWithUnitExtractor(PortugueseAgeExtractorConfiguration()),
                NumberWithUnitParser(PortugueseAgeParserConfiguration()))
        ]))
        # endregion

        # region Spanish
        self.register_model('CurrencyModel', Culture.Spanish, lambda options: CurrencyModel(
            [ExtractorParserModel(BaseMergedUnitExtractor(SpanishCurrencyExtractorConfiguration(
            )), BaseMergedUnitParser(SpanishCurrencyParserConfiguration()))]
        ))
        self.register_model('TemperatureModel', Culture.Spanish, lambda options: TemperatureModel(
            [ExtractorParserModel(NumberWithUnitExtractor(SpanishTemperatureExtractorConfiguration(
            )), NumberWithUnitParser(SpanishTemperatureParserConfiguration()))]
        ))
        self.register_model('DimensionModel', Culture.Spanish, lambda options: DimensionModel(
            [ExtractorParserModel(NumberWithUnitExtractor(SpanishDimensionExtractorConfiguration(
            )), NumberWithUnitParser(SpanishDimensionParserConfiguration()))]
        ))
        self.register_model('AgeModel', Culture.Spanish, lambda options: AgeModel(
            [ExtractorParserModel(NumberWithUnitExtractor(SpanishAgeExtractorConfiguration(
            )), NumberWithUnitParser(SpanishAgeParserConfiguration()))]
        ))
        # endregion

        # region Spanish Mexican
        self.register_model('CurrencyModel', Culture.SpanishMexican, lambda options: CurrencyModel(
            [ExtractorParserModel(BaseMergedUnitExtractor(SpanishCurrencyExtractorConfiguration(
            )), BaseMergedUnitParser(SpanishCurrencyParserConfiguration(culture_info=CultureInfo(Culture.SpanishMexican))))]
        ))
        self.register_model('TemperatureModel', Culture.SpanishMexican, lambda options: TemperatureModel(
            [ExtractorParserModel(NumberWithUnitExtractor(SpanishTemperatureExtractorConfiguration(
            )), NumberWithUnitParser(SpanishTemperatureParserConfiguration(culture_info=CultureInfo(Culture.SpanishMexican))))]
        ))
        self.register_model('DimensionModel', Culture.SpanishMexican, lambda options: DimensionModel(
            [ExtractorParserModel(NumberWithUnitExtractor(SpanishDimensionExtractorConfiguration(
            )), NumberWithUnitParser(SpanishDimensionParserConfiguration(culture_info=CultureInfo(Culture.SpanishMexican))))]
        ))
        self.register_model('AgeModel', Culture.SpanishMexican, lambda options: AgeModel(
            [ExtractorParserModel(NumberWithUnitExtractor(SpanishAgeExtractorConfiguration(
            )), NumberWithUnitParser(SpanishAgeParserConfiguration(culture_info=CultureInfo(Culture.SpanishMexican))))]
        ))
        # endregion

        # region Italian
        self.register_model('CurrencyModel', Culture.Italian, lambda options: CurrencyModel([
            ExtractorParserModel(
                BaseMergedUnitExtractor(
                    ItalianCurrencyExtractorConfiguration()),
                BaseMergedUnitParser(ItalianCurrencyParserConfiguration()))
        ]))
        # endregion

        # region German
        self.register_model('CurrencyModel', Culture.German, lambda options: CurrencyModel([
            ExtractorParserModel(
                BaseMergedUnitExtractor(
                    GermanCurrencyExtractorConfiguration()),
                BaseMergedUnitParser(GermanCurrencyParserConfiguration()))
        ]))
        # endregion

        # region Japanese
        self.register_model('CurrencyModel', Culture.Japanese, lambda options: CurrencyModel([
            ExtractorParserModel(
                BaseMergedUnitExtractor(
                    JapaneseCurrencyExtractorConfiguration()),
                BaseMergedUnitParser(JapaneseCurrencyParserConfiguration())),
            ExtractorParserModel(
                NumberWithUnitExtractor(
                    EnglishCurrencyExtractorConfiguration()),
                NumberWithUnitParser(EnglishCurrencyParserConfiguration()))
        ]))
        # endregion

    def get_age_model(self, culture: str = None, fallback_to_default_culture: bool = True) -> Model:
        return self.get_model('AgeModel', culture, fallback_to_default_culture)

    def get_currency_model(self, culture: str = None, fallback_to_default_culture: bool = True) -> Model:
        return self.get_model('CurrencyModel', culture, fallback_to_default_culture)

    def get_dimension_model(self, culture: str = None, fallback_to_default_culture: bool = True) -> Model:
        return self.get_model('DimensionModel', culture, fallback_to_default_culture)

    def get_temperature_model(self, culture: str = None, fallback_to_default_culture: bool = True) -> Model:
        return self.get_model('TemperatureModel', culture, fallback_to_default_culture)


def recognize_age(query: str, culture: str, options: NumberWithUnitOptions = NumberWithUnitOptions.NONE, fallback_to_default_culture: bool = True) -> List[ModelResult]:
    recognizer = NumberWithUnitRecognizer(culture, options)
    model = recognizer.get_age_model(culture, fallback_to_default_culture)
    return model.parse(query)


def recognize_currency(query: str, culture: str, options: NumberWithUnitOptions = NumberWithUnitOptions.NONE, fallback_to_default_culture: bool = True) -> List[ModelResult]:
    recognizer = NumberWithUnitRecognizer(culture, options)
    model = recognizer.get_currency_model(culture, fallback_to_default_culture)
    return model.parse(query)


def recognize_dimension(query: str, culture: str, options: NumberWithUnitOptions = NumberWithUnitOptions.NONE, fallback_to_default_culture: bool = True) -> List[ModelResult]:
    recognizer = NumberWithUnitRecognizer(culture, options)
    model = recognizer.get_dimension_model(
        culture, fallback_to_default_culture)
    return model.parse(query)


def recognize_temperature(query: str, culture: str, options: NumberWithUnitOptions = NumberWithUnitOptions.NONE, fallback_to_default_culture: bool = True) -> List[ModelResult]:
    recognizer = NumberWithUnitRecognizer(culture, options)
    model = recognizer.get_temperature_model(
        culture, fallback_to_default_culture)
    return model.parse(query)
